### DISCUSSION TITLE
`trigger` inside `useEffect`

### DISCUSSION THREAD

**Discussion by sellsjonathan** [react-hook-form/react-hook-form#8871] [Q&A] [answered]
I have an endless call loop happening inside a useEffect, where `trigger` is called.

In my situation, I have a parent button that needs to submit a file upload form, as well as a RHF controlled ChildForm.

Everything works as intended, except the validation requirements on the RHF child form.

I need to `trigger()` validation when the child form is submitted, in case there are some untouched `required` inputs.

The childform 'fake' submits itself based on a prop controlling a `useEffect` execution.

As soon as `trigger` is called within my `useEffect`, it endlessly forces a rerender, and doesn't let my submission boolean change from the parent `submitHandler`.

It can be fixed by calling `trigger` in a `setTimeout`, but I feel like this is kind of hacky. 

Am I missing something?

Here's a simplified demo of my issue :
https://codesandbox.io/s/xenodochial-varahamihira-ihmwo6?file=/src/ChildForm.tsx

**Comment by Moshyfawn** [comment]
I haven't looked into your use-case deeply, but from what I see, you might be able to avoid using `useEffect` all together, based on what you're trying to do _(I don't have the time check it out myself, sadly)_.

Check out [the `You Might Not Need an Effect` React docs article](https://beta.reactjs.org/learn/you-might-not-need-an-effect). Maybe it can help you

**Reply by sellsjonathan** [reply]
Hmm, technically you're right, there is no reason I couldn't do a boolean-ed `trigger()` call on every render based on the same dependency as I was keying off of in the useEffect.  

To me, a useEffect is the guard to reduce redundant or unneeded executions during a render.  

Do you think `trigger` should never be used in a useEffect then? If so surely this should get highlighted in the docs.

**Comment by sellsjonathan** [comment, marked as answer]
### Update:

So `trigger()` was forcing the rerender, which was actually recalculating a dependency that was not `useMemo`.  This was a calculation based on a value that came from a `useWatch`.  Adding a `useMemo` to that value stopped the runaway `useEffect`, but stopped getting the latest form values from the `useWatch`.


**Solution:**

```jsx
	useEffect(() => {
		const subscription = watch((data) => {
			setQuoteTotals(calculateQuoteTotals(data.quoteForm as IQuoteItem[]));
		});
		return () => {
			subscription.unsubscribe();
		};
	}, [watch]);

```
I ended up following the advanced guide in wiring up `watch` within a `useEffect`.  This updates on every change to the form values. Which allows me to get the latest user input to do my calculation in `calculateQuoteTotals` that is then saved to local state `setQuoteTotals`.

Unfortunately I still ended up with a runaway `useEffect`.

I ended up destructuring `isValid` from formState, so that it was a value as a dependency and not a reference, and everything is working as expected:


```jsx
	const { isValid } = formState;
	useEffect(() => {
		if (parentSubmitRequest) {
			trigger();
			submitHandler({
				totals: quoteTotals,
				quoteItems,
				isValid,
			});
		}
	}, [
		isValid,
		parentSubmitRequest,
		submitHandler,
		quoteItems,
		quoteTotals,
		trigger,
	]);

```
I will note that in the docs @bluebill1049 , you specifically say to add the entire `formState` as a dependency, which didn't work for me in this case. (https://react-hook-form.com/api/useform/formstate).

I'm mainly talking to myself, but maybe this will help someone in the future.

Cheers.
